src/Microsoft.TestPlatform.Extensions.TrxLogger/TrxLogger.cs

```diff
@@ -490,6 +456,199 @@ private void SetDefaultTrxFilePath()
+            testrunConfig.RunDeploymentRootDirectory = runDeploymentRoot;
+            this.testRun.RunConfiguration = testrunConfig;
+        }
+
+        /// <summary>
+        /// Gets test result from stored test results.
+        /// </summary>
+        /// <param name="executionId"></param>
+        /// <returns>Test result</returns>
+        private ITestResult GetTestResult(Guid executionId)
+        {
+            ITestResult testResult = null;
+
+            if (executionId != Guid.Empty)
+            {
+                this.results.TryGetValue(executionId, out testResult);
+
+                if (testResult == null)
+                    this.innerResults.TryGetValue(executionId, out testResult);
+            }
+
+            return testResult;
+        }
+
+        /// <summary>
+        /// Gets test element from stored test elements.
+        /// </summary>
+        /// <param name="testId"></param>
+        /// <returns></returns>
+        private ITestElement GetTestElement(Guid testId)
+        {
+            this.testElements.TryGetValue(testId, out var testElement);
+            return testElement;
+        }
+
+        /// <summary>
+        /// Gets or creates test element.
+        /// </summary>
+        /// <param name="executionId"></param>
+        /// <param name="parentExecutionId"></param>
+        /// <param name="testType"></param>
+        /// <param name="parentTestElement"></param>
+        /// <param name="rockSteadyTestCase"></param>
+        /// <returns>Trx test element</returns>
+        private ITestElement GetOrCreateTestElement(Guid executionId, Guid parentExecutionId, TestType testType, ITestElement parentTestElement, ObjectModel.TestCase rockSteadyTestCase)
+        {
+            ITestElement testElement = parentTestElement;
+            if (parentTestElement == null || parentTestElement.TestType.Equals(TrxLoggerConstants.OrderedTestType))
+            {
+                Guid testId = Converter.GetTestId(rockSteadyTestCase);
+                testElement = GetTestElement(testId);
+
+                if (testElement == null)
+                {
+                    testElement = Converter.ToTestElement(testId, executionId, parentExecutionId, testType, rockSteadyTestCase);
+                    testElements.Add(testId, testElement);
+                }
+            }
+
+            return testElement;
+        }
+
+        /// <summary>
+        /// Update test links
+        /// </summary>
+        /// <param name="testElement"></param>
+        /// <param name="parentTestElement"></param>
+        private void UpdateTestLinks(ITestElement testElement, ITestElement parentTestElement)
+        {
+            if (parentTestElement != null &&
+                parentTestElement.TestType.Equals(TrxLoggerConstants.OrderedTestType) &&
+                !(parentTestElement as OrderedTestElement).TestLinks.ContainsKey(testElement.Id.Id))
```

Comment:
In case orderedtest has data driven test, then we get same test element for all data driven tests. In that case we want to update test link only once.